flang/lib/Optimizer/Transforms/RewriteLoop.cpp

```diff
@@ -297,15 +297,28 @@ class CfgConversion : public CFGConversionBase<CfgConversion> {
   void runOnFunction() override {
     auto *context = &getContext();
     mlir::OwningRewritePatternList patterns(context);
-    patterns.insert<CfgLoopConv, CfgIfConv, CfgIterWhileConv>(
-        context, forceLoopToExecuteOnce);
     mlir::ConversionTarget target(*context);
     target.addLegalDialect<mlir::AffineDialect, FIROpsDialect,
                            mlir::arith::ArithmeticDialect,
                            mlir::StandardOpsDialect>();
 
     // apply the patterns
-    target.addIllegalOp<ResultOp, DoLoopOp, IfOp, IterWhileOp>();
+    if (transformWhileLoop) {
+      patterns.insert<CfgIterWhileConv>(context, forceLoopToExecuteOnce);
+      target.addIllegalOp<IterWhileOp>();
+    }
+    if (transformDoLoop) {
+      patterns.insert<CfgLoopConv>(context, forceLoopToExecuteOnce);
+      target.addIllegalOp<DoLoopOp>();
+    }
+    if (transformIf) {
+      patterns.insert<CfgIfConv>(context, forceLoopToExecuteOnce);
+      target.addIllegalOp<IfOp>();
+    }
+    bool transformAll = transformWhileLoop && transformDoLoop && transformIf;
+    if (transformAll) {
+      target.addIllegalOp<ResultOp>();
```

Comment:
Why is this illegal only if we enable all transformations?

Reply:
Because any of these ops can have a ResultOp. So, if they have not all been removed then it is likely there is also some ResultOps around.

Reply:
Right makes sense